Sources/Core/Sources/ECS/Systems/PlayerInputSystem.swift

```diff
@@ -136,16 +136,60 @@ public final class PlayerInputSystem: System {
       if event.key == .enter {
         if !message.isEmpty {
           try connection?.sendPacket(ChatMessageServerboundPacket(message))
+          guiState.playerMessageHistory.append(message)
+          guiState.currentMessageIndex = nil
         }
         guiState.messageInput = nil
         return true
       } else if event.key == .escape {
         guiState.messageInput = nil
+        guiState.currentMessageIndex = nil
         return true
       } else if event.key == .delete {
         if !message.isEmpty {
           guiState.messageInput?.removeLast()
         }
+      } else if event.key == .upArrow {
+        // If no message is selected, select the above message
+        if guiState.currentMessageIndex == nil {
+          guiState.stashedMessageInput = guiState.messageInput
+          // If there are no messages, do nothing
+          if guiState.playerMessageHistory.count > 0 {
+            // Index up a message
+            guiState.currentMessageIndex = guiState.playerMessageHistory.count - 1
+            if let currentMessageIndex: Int = guiState.currentMessageIndex {
+              // Set the message input to the message at the index
+              guiState.messageInput = guiState.playerMessageHistory[guiState.currentMessageIndex!]
+            }
+          }
+        } else {
+          // If there is a message selected, index up a message
+          if guiState.currentMessageIndex! > 0 {
+            guiState.currentMessageIndex! -= 1
+          }
+          if let currentMessageIndex: Int = guiState.currentMessageIndex {
+            // Set the message input to the message at the index
+            guiState.messageInput = guiState.playerMessageHistory[guiState.currentMessageIndex!]
+          }
+        }
+      } else if event.key == .downArrow {
+        // If there is a message selected, index down a message
+        if guiState.currentMessageIndex != nil {
+          if guiState.currentMessageIndex! < guiState.playerMessageHistory.count - 1 {
+            guiState.currentMessageIndex! += 1
+            if let currentMessageIndex: Int = guiState.currentMessageIndex {
```

Comment:
`if let` is only useful if you use `currentMessageIndex` in the body of the if statement. In this case you don't and you force unwrap anyway. With the change I proposed for line 177 you can get rid of the if statement entirely, but it's just something to keep in mind.